test-e2e/test_master_node_replacement.py

```diff
@@ -0,0 +1,209 @@
+    ensemble, after removing one master, and then adding another master with
+    the same IP address, the cluster will get to a healthy state. This is
+    repeated until all masters in the original cluster have been replaced.
+    The purpose of this test is to assert that the ``node-poststart``
+    procedure correctly prevents a master node replacement from being performed
+    too quickly. A new master node should only become part of the cluster if
+    there are no more underreplicated ranges reported by CockroachDB.
+
+    Permanent CockroachDB data loss and a potential breakage of DC/OS occurs
+    when a second master node is taken down for replacement while CockroachDB
+    is recovering and there are still underreplicated ranges due to a recent
+    other master node replacement.
+    """
+    docker_backend = Docker(network=docker_network_three_available_addresses)
+
+    with Cluster(
+        cluster_backend=docker_backend,
+        # Allocate all 3 available IP addresses in the subnet.
+        masters=3,
+        agents=0,
+        public_agents=0,
+    ) as original_cluster:
+        master = next(iter(original_cluster.masters))
+        result = master.run(
+            args=[
+                'ifconfig',
+                '|', 'grep', '-B1', str(master.public_ip_address),
+                '|', 'grep', '-o', '"^\w*"',
+            ],
+            shell=True,
+        )
+        interface = result.stdout.strip().decode()
+        ip_detect_contents = textwrap.dedent(
+            """\
+            #!/bin/bash -e
+            if [ -f /sbin/ip ]; then
+               IP_CMD=/sbin/ip
+            else
+               IP_CMD=/bin/ip
+            fi
+
+            $IP_CMD -4 -o addr show dev {interface} | awk '{{split($4,a,"/");print a[1]}}'
+            """.format(interface=interface),
+        )
+        ip_detect_path = tmp_path / 'ip-detect'
+        ip_detect_path.write_text(data=ip_detect_contents)
+        static_config = {
+            'master_discovery': 'static',
+            'master_list': [str(master.private_ip_address)
+                            for master in original_cluster.masters],
+        }
+        dcos_config = {
+            **original_cluster.base_config,
+            **static_config,
+        }
+        original_cluster.install_dcos_from_path(
+            dcos_installer=artifact_path,
+            dcos_config=dcos_config,
+            ip_detect_path=ip_detect_path,
+        )
+        wait_for_dcos_oss(
+            cluster=original_cluster,
+            request=request,
+            log_dir=log_dir,
+        )
+        current_cluster = original_cluster
+        tmp_clusters = set([])
+
+        original_masters = original_cluster.masters
+
+        try:
+            for master_to_be_replaced in original_masters:
+                # Destroy a master and free one IP address.
+                original_cluster.destroy_node(node=master_to_be_replaced)
```

Comment:
You could use `current_cluster` here, right? I think that would be clearer, as, conceptually, the original cluster no longer exists (at least after the first iteration).